lib/galaxy/model/__init__.py

```diff
@@ -1455,7 +1455,9 @@ def __add_datasets_optimized(self, datasets, genome_build=None):
         set_genome = genome_build not in [None, '?']
         for i, dataset in enumerate(datasets):
             dataset.hid = base_hid + i
-            dataset.history = self
+            # Don't let SA manage this.
+            delattr(dataset, "history")
```

Comment:
Do you know why this speeds up things ?
Is it possible this prevents an HDA from being marked dirty, which results in a HDA copy ?
I've been looking at unnecessary HDA version bumps, I suspect something like
```
--- a/lib/galaxy/model/base.py
+++ b/lib/galaxy/model/base.py
@@ -7,8 +7,12 @@ from inspect import (
     isclass
 )

-from sqlalchemy import event
+from sqlalchemy import (
+    event,
+    inspect
+)
 from sqlalchemy.orm import (
+    class_mapper,
     scoped_session,
     sessionmaker
 )
@@ -52,7 +56,20 @@ class ModelMapping(Bunch):
 def versioned_objects(iter):
     for obj in iter:
         if hasattr(obj, '__create_version__'):
-            yield obj
+            if object_has_changed(obj):
+                yield obj
+
+
+def object_has_changed(myobj):
+    """We check wether the objbect itself has changed."""
+
+    inspr = inspect(myobj)
+    attrs = class_mapper(myobj.__class__).column_attrs  # exclude relationships
+    for attr in attrs:
+        hist = getattr(inspr.attrs, attr.key).history
+        if hist.has_changes():
+            return True
+    return False
```
might result in a similar speedup ?

Reply:
That isn't a speed up - the speed up is setting ``history_id`` instead of ``history`` - the way I do other places (https://github.com/galaxyproject/galaxy/pull/6524/files#diff-bfb557a99c1f7d646d4968d8d680b885R645). Usually that trick just seems to work, here it didn't because the dataset already had dataset.history = None set on it in its constructor. So SA was opting to respect the null history instead of using the history_id that I was trying to set.